File: TCPServer/Fortran/TCPServer.f90
program tcpserver
 use iso_c_binding
 implicit none
 interface ! These are basic C bindings for fortran. Look up the fortran FFI online for more info
  integer(c_int) function socket(family,type,protocol) bind(c,name='socket')
   import::c_int
   integer(c_int),value,intent(in)::family,type,protocol
   end function socket
  integer(c_int) function setsockopt(socket,layer,option,value,size) bind(c,name='setsockopt')
   import::c_int,c_ptr
   integer(c_int),value,intent(in)::socket,layer,option,size
   type(c_ptr),value,intent(in)::value
   end function setsockopt
  integer(c_int) function bind(socket,address,length) bind(c,name='bind')
   import::c_int,c_ptr
   integer(c_int),value,intent(in)::socket,length
   type(c_ptr),value,intent(in)::address
   end function bind
  integer(c_int) function listen(socket,backlog) bind(c,name='listen')
   import::c_int
   integer(c_int),value,intent(in)::socket,backlog
   end function listen
  integer(c_int) function accept(socket,address,addrsize) bind(c,name='accept')
   import::c_int,c_ptr
   integer(c_int),value,intent(in)::socket
   type(c_ptr),value,intent(in)::address,addrsize
   end function accept
  integer(c_int) function read(fd,buffer,size) bind(c,name='read')
   import::c_int,c_ptr
   integer(c_int),value,intent(in)::fd,size
   type(c_ptr),value,intent(in)::buffer
   end function read
  integer(c_int) function write(fd,buffer,size) bind(c,name='write')
   import::c_int,c_ptr
   integer(c_int),value,intent(in)::fd,size
   type(c_ptr),value,intent(in)::buffer
   end function write
  integer(c_int) function close(fd) bind(c,name='close')
   import::c_int
   integer(c_int),value,intent(in)::fd
   end function close
  end interface
 integer(c_int8_t),dimension(256),target::buffer
 character(len=17),target::reply="Message received"//new_line('1')
 type::sockaddr
  integer(c_short)::type=2                     ! Address family of 2 is AF_INET/IPV4
  integer(c_int8_t)::porthi=SHIFTR(3721,8)     ! Port low high byte
  integer(c_int8_t)::portlo=IAND(3721,255)     ! Port low byte
  integer(c_int8_t),dimension(4)::ip=[0,0,0,0] ! IP address. 0.0.0.0 allows any ip to connect
  integer(c_int8_t),dimension(8)::pad          ! Padding
  end type sockaddr
 type(sockaddr),target::address
 integer(c_int)::server,client,result,size
 integer(c_int),target::true=1,addrsize=16
 server=socket(2,1,0)                          ! Create the server
 result=setsockopt(server,1,15,c_loc(true),4)  ! Set reuse for quick restarts
 result=bind(server,c_loc(address),addrsize)   ! Bind the address
 result=listen(server, 5)                      ! Set the socket as listener
 do while(1==1)
  client=accept(server,c_loc(address),c_loc(addrsize)) ! Accept a client
  size=read(client,c_loc(buffer),255)                  ! Read a message from the client
  size=write(1,c_loc(buffer),size)                     ! Write the message to stdout
  size=write(client,c_loc(reply),17)                   ! Send the response to the client
  result=close(client)                                 ! Close the client
  end do
 end program tcpserver

File: TCPServer/Haskell/TCPServer.hs
module Main where
 import Data.Bits
 import Foreign.C
 import Foreign.C.Types
 import Foreign.Ptr
 import Foreign.Storable
 import Foreign.Marshal.Alloc
 foreign import ccall "socket"     c_socket::CInt->CInt->CInt->IO CInt
 foreign import ccall "setsockopt" c_setsockopt::CInt->CInt->CInt->Ptr CInt->CInt->IO CInt
 foreign import ccall "bind"       c_bind::CInt->Ptr a->CInt->IO CInt
 foreign import ccall "listen"     c_listen::CInt->CInt->IO CInt
 foreign import ccall "accept"     c_accept::CInt->Ptr a->Ptr CInt->IO CInt
 foreign import ccall "read"       c_read::CInt->Ptr a->CInt->IO CInt
 foreign import ccall "write"      c_write::CInt->Ptr a->CInt->IO CInt
 foreign import ccall "close"      c_close::CInt->IO CInt
 main=do
  true<-mallocBytes 4
  addrsize<-mallocBytes 4
  addr<-mallocBytes 16
  buffer<-mallocBytes 256
  poke (true::Ptr CInt) 1
  poke (addr::Ptr CShort) 2                             --Address family 2, AF_INET/IPV4
  poke ((plusPtr addr 2)::Ptr CUShort) (shiftR 3721 8)  --Port high byte. Have to CUShort because because the value is cast before shifting
  poke ((plusPtr addr 3)::Ptr CUChar) ((.&.) 3721 0xFF) --Port low byte. The ip and padding should be initialized to 0. IP 0.0.0.0 accepts connections from any address.
  server<-c_socket 2 1 0                  --Create server socket
  result<-c_setsockopt server 1 15 true 4 --Set reuse for quick restarts
  result<-c_bind server addr 16           --Bind address
  result<-c_listen server 5               --Set socket as listener
  mainloop server addr addrsize buffer
 mainloop server addr addrsize buffer=do
  reply<-newCString "Message recieved\n"
  client<-c_accept server addr addrsize   --Accept client
  size<-c_read client buffer 0xFF         --Read message from client
  result<-c_write 1 buffer size           --Write message to stdout
  result<-c_write client reply 17         --Send reply to client
  result<-c_close client                  --Close client
  mainloop server addr addrsize buffer